### PR TITLE
mtmd : support Qwen 2.5 Omni (input audio+vision, no audio output)

### PR DESCRIPTION
**Pull request by ngxson**
This PR aimed to just add the capability to use 2 `clip_ctx`, one for audio and one for vision. But I ended up doing quite more refactoring than I initially thought

- [x] some fields are moved from `clip_ctx` to `clip_model`
- [x] from a single `clip_model_loader`, we are now able to create multiple `clip_model` and `clip_ctx`
- [x] `libmtmd` can handle 2 `clip_ctx` and switch the calls according to the chunk type
- [x] refactor `mtmd_tokenize` so it can handle mixed modality
- [x] nits: some functions in `clip.cpp` has many `if..else` branches, we can refactor them to `switch (...)`
- [x] refactor tests, it can now support testing audio input
- [x] implement `SinusoidsPositionEmbedding` used by the audio encoder --> generate it during conversion
- [x] fix M-RoPE position for audio chunk

TODO in next PRs:
- Remove unused clip APIs
- Move image preprocessing to `mtmd-image.cpp`
- Move file reader/decoder to `mtmd-helper.cpp`

## Why no audio output?

The simple answer is: I don't have time to implement it.

The long answer: Qwen 2.5 Omni generates audio using 2 steps:
- Step 1: generate the mel spectrogram using DiT (diffusion transformer). This is essentially image generation, so kinda give-away what Qwen is cooking next 😃 
- Step 2: turn mel into wav using BigVGAN

So adding audio generation is indeed adding image generation capability, which I don't really have time to do right now.

## Demo

Pre-quantized models:

```sh
# Qwen2.5 Omni
# Capabilities: audio input, vision input
(tool_name) -hf ggml-org/Qwen2.5-Omni-3B-GGUF
(tool_name) -hf ggml-org/Qwen2.5-Omni-7B-GGUF
```

image

### CODE REVIEW

**Comment by ngxson** [comment]
@ggerganov Sorry for including quite a lot of changes in one single PR.

The global idea of this PR is to use 2 dedicated `clip_ctx`, each for one modality. I also thought of maybe having audio output and image output support in the future, so this model of multiple `clip_ctx` can be useful to make it happen.

A given `clip_model` is attached to one `clip_ctx` for now (so multiple context cannot use the same model). As each modality has its own set of tensors, so I don't feel like it's necessary to split model & ctx like on `libllama`. However, only one `model_loader` is used on reading the model. I think it's more useful this way because it can tell in advance how many contexts should be created (and also, model file is only opened once)

On `libmtmd` side, I rewrite the `tokenize` function to handle mixed image/audio chunks. It can also handle some edge cases (cc @mattjcly ), for example:
- Empty input
- Input with only one media, no text

**Comment by mega-cqz** [comment]
Amazing work!!

**Review by ggerganov — APPROVED** [review]

File: tools/mtmd/clip.cpp [hunk 1]
```diff
@@ -2554,18 +2588,18 @@ struct clip_model_loader {
...
-        if (hparams.has_vision) {
+        if (ctx_clip.model.modality == CLIP_MODALITY_VISION) {
             img->nx = hparams.warmup_image_size;
             img->ny = hparams.warmup_image_size;
         } else {
-            img->nx = 1024; // TODO @ngxson : use a better default
+            img->nx = hparams.warmup_audio_size;
             img->ny = hparams.n_mel_bins;
         }
         img->buf.resize(img->nx * img->ny * 3);
```
[the remark this answers is not in the document]
Comment (ngxson) [reply tools/mtmd/clip.cpp]: Indeed, only the image shape is needed during this warmup, so we don't actually need to allocate this buffer. I removed it in https://github.com/ggml-org/llama.cpp/pull/13784/commits/05310968743ab1777719eb3e3f6f28fafc8fa620

File: tools/mtmd/mtmd-helper.cpp [hunk 2]
```diff
@@ -85,6 +86,23 @@ struct decode_embd_batch {
...
+            pos[i + batch.n_tokens * 2] = pos_0 + i;
+            pos[i + batch.n_tokens * 3] = 0; // last pos dim is unused
+        }
+        for (int i = 0; i < batch.n_tokens; i++) {
+            batch.n_seq_id[i] = 1;
+            batch.seq_id  [i] = seq_id_0.data();
+            batch.logits  [i] = false;
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (ngxson) [reply tools/mtmd/mtmd-helper.cpp]: Yes thanks for noticing, it's a leftover code from the 2D version, I removed it in https://github.com/ggml-org/llama.cpp/pull/13784/commits/27a8f26685e61e039e0c18aa987d94d980ef0ef4

**Comment by henfiber** [comment]
Just fyi, with Vulkan backend (AMD RADV RENOIR) and the Q8_0 mmproj file the server crashes with:  

> Floating point exception (core dumped)  

<details><summary>Details</summary>
<pre>
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = AMD Radeon Graphics (RADV RENOIR) (radv) | uma: 1 | fp16: 1 | warp size: 64 | shared memory: 65536 | int dot: 0 | matrix cores: none
build: 5510 (a8ea03d8) with cc (GCC) 14.2.1 20250110 (Red Hat 14.2.1-7) for x86_64-redhat-linux
system info: n_threads = 5, n_threads_batch = 6, total_threads = 12

system_info: n_threads = 5 (n_threads_batch = 6) / 12 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

clip_model_loader: model name:   
clip_model_loader: description:  
clip_model_loader: GGUF version: 3
clip_model_loader: alignment:    32
clip_model_loader: n_tensors:    1008
clip_model_loader: n_kv:         32

clip_model_loader: has vision encoder
clip_model_loader: has audio encoder
clip_ctx: CLIP using Vulkan0 backend
load_hparams: projector:          qwen2.5o
load_hparams: n_embd:             1280
load_hparams: n_head:             16
load_hparams: n_ff:               1280
load_hparams: n_layer:            32
load_hparams: ffn_op:             silu
load_hparams: projection_dim:     3584

--- vision hparams ---
load_hparams: image_size:         1024
load_hparams: patch_size:         14
load_hparams: has_llava_proj:     0
load_hparams: minicpmv_version:   0
load_hparams: proj_scale_factor:  0
load_hparams: n_wa_pattern:       8

load_hparams: model size:         1476.70 MiB
load_hparams: metadata size:      0.35 MiB
alloc_compute_meta:    Vulkan0 compute buffer size =     2.77 MiB
alloc_compute_meta:        CPU compute buffer size =     0.16 MiB
clip_ctx: CLIP using Vulkan0 backend
load_hparams: projector:          qwen2.5o
load_hparams: n_embd:             1280
load_hparams: n_head:             20
load_hparams: n_ff:               5120
load_hparams: n_layer:            32
load_hparams: ffn_op:             gelu_erf
load_hparams: projection_dim:     3584

--- audio hparams ---
load_hparams: n_mel_bins:         128
load_hparams: proj_stack_factor:  0

load_hparams: model size:         1476.70 MiB
load_hparams: metadata size:      0.35 MiB
Floating point exception (core dumped)
</pre>
</details> 

(mmproj-Qwen2.5-Omni-7B-Q8_0.gguf)

The error is avoided with `--no-mmproj-offload`.  

The FP16 mmproj file works properly.  

Maybe, there is an issue with the Q8_0 file or with the way the Vulkan backend handles the mixed precision format. Other Q8_0 mmproj files for other models (e.g. mmproj-InternVL3-2B-Instruct-Q8_0.gguf)  work properly (although this is a vision-only model). 

(Did not open an issue in case the issue is with the file and not the backend).

**Comment by pwilkin** [comment]
@ngxson Just out of curiosity, how much work is it to implement diffusion support? It seems like there are more and more models coming out with "all-to-all" capabilities (like Bagel), would be probably nice for image generation to appear on the roadmap at some point...

**Comment by nqchieutb01** [comment]
@ngxson 
I've just fine-tuned the Qwen2.5-Omni-7B model using LoRA, and then merged the adapters into the base model. Now, I’d like to convert this merged model to the GGUF format to run it with llama.cpp.
Could you please guide me on how to do this?

Thanks in advance!

**Comment by alielmorsy** [comment]
So, Will we ever have support for videos?

**Comment by nguyen9x** [comment]
> Vậy, liệu chúng ta có bao giờ được hỗ trợ video không?

Have you tested the entire omni function in this model?
